src/main/java/com/mongodb/jdbc/MongoSQLResultSetMetaData.java

```diff
@@ -2,57 +2,106 @@
 
 import java.sql.ResultSetMetaData;
 import java.sql.SQLException;
-import java.sql.SQLFeatureNotSupportedException;
-import org.bson.BsonType;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
 
 public class MongoSQLResultSetMetaData extends MongoResultSetMetaData implements ResultSetMetaData {
+    private static class NameSpace {
+        String datasource;
+        String columnLabel;
 
-    public MongoSQLResultSetMetaData(MongoResultDoc metadataDoc) {
-        super(metadataDoc);
+        NameSpace(String datasource, String columnLabel) {
+            this.datasource = datasource;
+            this.columnLabel = columnLabel;
+        }
     }
 
-    @Override
-    public int getColumnCount() throws SQLException {
-        throw new SQLFeatureNotSupportedException("TODO");
-    }
+    private static class DatasourceAndIndex {
+        String datasource;
+        int index;
 
-    @Override
-    public int isNullable(int column) throws SQLException {
-        throw new SQLFeatureNotSupportedException("TODO");
+        DatasourceAndIndex(String datasource, int index) {
+            this.datasource = datasource;
+            this.index = index;
+        }
     }
 
-    @Override
-    public String getColumnLabel(int column) throws SQLException {
-        throw new SQLFeatureNotSupportedException("TODO");
+    // A mapping from columnLabel name to datasource name and index.
+    private Map<String, DatasourceAndIndex> columnLabels;
+    // A mapping from index position to NameSpace (datasource, columnLabel).
+    private List<NameSpace> columnIndices;
+    // A mapping from index position to ColumnTypeInfo.
+    private List<MongoSQLColumnInfo> columnInfo;
+
+    // This gets the datasource for a given columnLabel, and is used
+    // in MongoSQLResultSet to retrieve data by label.
+    String getDatasource(String columnLabel) {
+        return columnLabels.get(columnLabel).datasource;
     }
 
-    @Override
-    public String getColumnName(int column) throws SQLException {
-        throw new SQLFeatureNotSupportedException("TODO");
+    private void assertObjectSchema(MongoJsonSchema schema) throws SQLException {
+        if (!schema.isObject() || schema.properties == null) {
```

Comment:
```suggestion
        if (!schema.isObject()) {
```
Should we also get rid of the `schema.properties == null` here and depend on `schema.isObject()` to capture the correct definition of "object schema"?

Reply:
I'll add a comment clarifying this, too